Feat/new blocksnet version

Voronapxl opened the pull request:
1. Refactor of effects_service.py
2. Addition of new context and scenario classes
3. Logic for new F22
4. Caching logic for context F22
5. Added response schemas to all methods
6. Added docstring to all methods
7. Deleted unnecessary GET methods, now all calculations are performed via POST task routers

Voronapxl committed:
feat(f_22): new f 22

Voronapxl committed:
fat(f_22):
1. Working logic for F 22

Voronapxl committed:
feat(f_22): tmp_logic

Voronapxl committed:
fat(f_22):
1. Additional calculations for project scenarios added - each scenario is now calculated

Voronapxl committed:
feat(effects_logic):
1. Refactor effects_service.py
2. Caching logic for context layer in pkl, parquet
3. Context service is now a class and some of the effects_service.py has moved into it
4.Scenario service is now a class and some of the effects_service.py has moved into it

Voronapxl committed:
feat(f22):
1. logic for tasks in F 22

Voronapxl committed:
feat(effects_logic): small fixes
1. Rewrite graph creation
2. Values transformation services sorting
3. F 22 tasks addition
4. Invalid parameters fix

Voronapxl committed:
feat(f22):
1. Converted response for F22
2. Started to implement response schemas
3. Added docstring for endpoints

Voronapxl committed:
fix(f36):
1. Logic for when child scenario is provided, calculations for base scenario is performed (only!)
2. Tasks addition for values_oriented_requirements
3. Response schema and description for get_service_types router

Voronapxl committed:
fix(f35):
1. Implementation of response schema for F35 territory_transformation

Voronapxl committed:
feat(response_schema):
1. Added response schemas for all endpoints

Voronapxl committed:
fix(response_schema):
1. Fix for SocioEconomicMetricsResponseSchema
2. Round of social_economical_effects

Voronapxl committed:
fix(main):
1. Return of exceptions

soldey commented on app/common/utils/effects_utils.py:
```diff
@@ -0,0 +1,164 @@
+from app.clients.urban_api_client import UrbanAPIClient
+
+
+class EffectsUtils:
```
разнести методы отсюда по сервисам

Voronapxl replied:
Вернул методы отсюда в EffectsService, если они использовались только там

soldey commented on app/effects_api/tasks_controller.py:
```diff
@@ -110,8 +139,81 @@ async def create_task(
+    _task_map[task_id] = task
+    await _task_queue.put(task)
+
+    return {"task_id": task_id, "status": "queued"}
```
унести логику в task_service

soldey reviewed (changes requested):
Я бы ещё добавил какие-то try except, чтобы не просто 500 ловить, а понимать хотя бы куски кода, в которых что-то падает (ну и вместе с этим можно logger.exception прокидывать на ошибку)

kanootoko commented on app/clients/urban_api_client.py:
```diff
@@ -401,3 +397,9 @@ async def get_project_scenarios(
         res = await self.json_handler.get("/api/v1/social_values")
         return {item["soc_value_id"]: item["name"] for item in res}
+
+    async def get_territory_normatives(self, territory_id: int):
```
А тут потерялась аннотация возвращаемого значения

kanootoko commented on app/common/utils/effects_utils.py:
```diff
@@ -0,0 +1,164 @@
+        p = s.get("parent_scenario")
+        return p.get("id") if isinstance(p, dict) else p
+
+    def sid(self, s: Dict[str, Any]) -> Optional[int]:
```
Это сокращение `service_id` до `sid` правда оправдано?

kanootoko commented on app/common/utils/effects_utils.py:
```diff
@@ -0,0 +1,164 @@
+        self,
+        urban_api_client: UrbanAPIClient,
+    ):
+        self.__name__ = "EffectsUtils"
```
Можно в целом и в атрибут класса вынести, раз он один для всех экземпляров

kanootoko commented on app/common/utils/effects_utils.py:
```diff
@@ -0,0 +1,164 @@
+        )
+        return self.sid(matches[0]) or scenario_id
+
+    def clean_number(self, v):
```
Я бы к таким штукам докстрингу добавлял, или примеры использования. Не с первого взгляда становится ясно, что оно пытается ~распарсить~ преобразовать значение из np-классов во float или int.

Ну, и наличие self в методах, которые его не используют, только путает

kanootoko commented on app/common/utils/geodata.py:
```diff
@@ -44,7 +44,7 @@ def safe_gdf_to_geojson(
     - Return parsed dict (FeatureCollection).
     """
-    logger.info("Serializing GeoDataFrame to GeoJSON (EPSG:%s, round=%d)", to_epsg, round_ndigits)
+    logger.info(f"Serializing GeoDataFrame to GeoJSON (EPSG:{to_epsg}, round={round_ndigits})")
```
Не-не, оно было лучше как раз так - разве что вместо %s и %d логгер вполне себе поддерживает `{}` для format'а.

Идея в том, что когда ты используешь f-строку, ты в любом случае совершаешь подстановку. Логгер же вызовет `.format` только если лог должен быть выведен (логгирование заданного уровня включено)

Voronapxl replied:
Я попробовал, у меня не получилось с ходу, поэтому вернулся к старому доброму f стрингу

kanootoko commented on app/common/utils/geodata.py:
```diff
@@ -128,4 +128,32 @@ def _ensure_block_index(gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+  return dist_mx // SPEED
+
+
+async def _roads_overlay_fast(self, scenario_id: int, token: str, target_crs, bounds):
```
Это что такое? Почему это функция (а не метод), и у нее первый параметр называется `self`? Не надо так путать

Voronapxl replied:
А это чудо я забыл убрать из кода

kanootoko commented on app/effects_api/constants/const.py:
```diff
@@ -18,70 +21,96 @@
+    31 : None, # фельдшерско-акушерский пункт
+    32 : None, # женская консультация
+    34 : 'pharmacy',
+    35 : 'hospital',
```
Если вернуть двойные кавычки, то половина diff'а уйдет. Ну так, если разницы нет

Voronapxl replied:
Скопировано с одного из обновлённых экзамплов, так что разницы в нет

kanootoko replied:
(как же я ошибался по поводу diff'а)

kanootoko commented on app/effects_api/modules/context_service.py:
```diff
@@ -1,129 +1,347 @@
+    """Context layer orchestration (blocks, buildings, services, fzones)."""
+
+    def __init__(self, urban_api_client: UrbanAPIClient, cache: FileCache
+        ):
```
Я не настаиваю на использовании авто-форматирования, но когда вижу такие штуки (закрывающие скобки, для которых не выполняется ни одно логическо обоснованное правило, как найти открывающую), `):` - это мое выражение лица

kanootoko commented on app/effects_api/modules/service_type_service.py:
```diff
@@ -147,3 +154,233 @@ async def remap_properties_keys_in_geojson(
+            suffix = col[len(pref):]
+
+            if _NUM_SUFFIX_RE.match(suffix):
+                break
```
Это ведь... `if suffis.isnumeric()`

kanootoko commented on app/effects_api/modules/services_service.py:
```diff
@@ -44,3 +46,51 @@ def adapt_services(
         for st in sorted(gdf["service_type"].unique())
     }
+
+async def get_services_layer(self, scenario_id: int, token: str):
```
Еще одна функция, принимающая self и без аннотаций типов возврата

Voronapxl replied:
Вообще нигде не использовалось, убрал

kanootoko commented on app/effects_api/schemas/socio_economic_metrics_response_schema.py:
```diff
@@ -0,0 +1,7 @@
+
+class SocioEconomicMetricsResponseSchema(BaseModel):
+    results: dict[str, dict[str, dict[str, int | float]]]
+    # results: dict
```
упс

kanootoko commented on app/effects_api/schemas/territory_transformation_response_schema.py:
```diff
@@ -0,0 +1,29 @@
+from app.common.dto.models import FeatureCollectionModel
+
+
+# from pydantic_geojson import FeatureCollectionModel
```
Ну и это тоже

kanootoko commented on app/effects_api/schemas/values_tables_response_schema.py:
```diff
@@ -0,0 +1,15 @@
+    )
+    services_type_deficit: List[Dict[str, str | float | List[str]]] = Field(
+        None, description="Services type deficit for the base scenario"
+    )
```
А None здесь ведь является дефолтным значением, разве нет? При этом аннотации не предполагают Optional'а. (Могу ошибаться, но обратил внимания из-за отсутствия `...` первым аргументом)

Voronapxl committed:
fix(comments_tmp):
1. Comment fixes

Voronapxl committed:
fix(comment_fixes):
1. black formating
2. fix for optional values in response schemas
3. Deleted redundant functions

Voronapxl committed:
fix(comment_fixes):
1. moved logic from controller to task service

soldey commented on app/effects_api/effects_service.py:
```diff
@@ -1046,19 +1223,71 @@ async def _compute_for_single_scenario(
+        """
+        Transform scenario-first results to territory-first pivot.
+
+        Input:
+            results: {
+                scenario_id: [
+                    {"territory_id": int, "indicator_id": int, "value": number},
+                    ...
+                ],
+                ...
+            }
```
на такой случай надо прописывать dto

Voronapxl committed:
fix(comment_fixes):
1. Small additions in endpoint descriptions
